app/src/main/java/com/futurice/freesound/feature/common/ui/adapter/RecyclerViewAdapter.java

```diff
@@ -0,0 +1,158 @@
+        final DisplayableItem item = modelItems.get(position);
+        binderMap.get(item.type()).bind(get(holder), item);
+    }
+
+    @Override
+    public void onViewRecycled(final RecyclerView.ViewHolder holder) {
+        super.onViewRecycled(holder);
+        if (holder instanceof BaseBindingViewHolder) {
+            ((BaseBindingViewHolder) holder).unbind();
+        }
+    }
+
+    @Override
+    public int getItemCount() {
+        return modelItems.size();
+    }
+
+    @Override
+    public int getItemViewType(final int position) {
+        return modelItems.get(position).type();
+    }
+
+    /**
+     * Updates modelItems currently stored in adapter with the new modelItems.
+     *
+     * @param items collection to update the previous values
+     */
+    public void update(@NonNull final List<DisplayableItem> items) {
+        Preconditions.checkNotNull(items);
+        assertUiThread();
+
+        if (modelItems.isEmpty()) {
+            updateAllItems(items);
+        } else {
+            updateDiffItemsOnly(items);
+        }
+    }
+
+    /**
+     * Only use for the first update of the adapter, whe it is still empty.
+     */
+    private void updateAllItems(@NonNull final List<DisplayableItem> items) {
+        Observable.just(items)
+                  .doOnNext(this::updateItemsInModel)
+                  .subscribe(__ -> notifyDataSetChanged(),
+                             e -> Timber.e(e, "Error notifying data set changed in adapter"));
+    }
+
+    /**
+     * Do not use for first update of the adapter.
+     * The method {@link DiffUtil.DiffResult#dispatchUpdatesTo(RecyclerView.Adapter)} is
+     * significantly slower than {@link RecyclerViewAdapter#notifyDataSetChanged()} when it comes
+     * to update all the items in the adapter.
+     */
+    private void updateDiffItemsOnly(@NonNull final List<DisplayableItem> items) {
+        final List<DisplayableItem> itemsCopy = new ArrayList<>(items);
+        Observable.fromCallable(() -> calculateDiff(itemsCopy))
+                  .subscribeOn(schedulerProvider.computation())
+                  .observeOn(schedulerProvider.ui())
+                  .doOnNext(__ -> updateItemsInModel(items))
+                  .subscribe(this::updateAdapterWithDiffResult,
+                             e -> Timber.e(e, "Error dispatching updates to adapter"));
+    }
+
+    private DiffUtil.DiffResult calculateDiff(@NonNull final List<DisplayableItem> newItems) {
+        return DiffUtil.calculateDiff(new DiffUtilCallback(modelItems, newItems, comparator));
+    }
+
+    private void updateItemsInModel(@NonNull final List<DisplayableItem> items) {
+        modelItems.clear();
+        modelItems.addAll(items);
+    }
+
+    private void updateAdapterWithDiffResult(@NonNull final DiffUtil.DiffResult result) {
+        result.dispatchUpdatesTo(this);
+    }
+
+    private void assertUiThread() {
```

Comment:
What do you think about having this method in the `SchedulerProvider`? I can see it being useful for reuse.